scilla/src/call_scilla_server.rs

```diff
@@ -0,0 +1,291 @@
+use std::{
+    io::{Read, Write},
+    net::TcpStream,
+    str,
+};
+
+use std::path::Path;
+use std::fs::File;
+use anyhow::{anyhow, Result};
+use jsonrpc_core::Params;
+use serde::{Deserialize, Serialize};
+use serde_json::{from_str, Value};
+use tracing::*;
+
+use crate::scilla_tcp_server::ScillaServer;
+
+/// Collection of functions to call the Scilla server and decode the result.
+/// The communications are over TCP currently.
+/// The scilla server is unusual in that it sometimes responds with a string as the 'result' which
+/// then has to be further parsed.
+/// There are two tcp connections required. The first is for the request (such as a 'run' command),
+/// and the second is for the backend queries (the server reading and writing to the state).
+
+/// Response from the 'check' command
+#[derive(Deserialize, Debug)]
+pub struct CheckOutput {
+    pub contract_info: ContractInfo,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct ContractInfo {
+    pub scilla_major_version: String,
+    pub fields: Vec<Param>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Param {
+    #[serde(rename = "vname")]
+    pub name: String,
+    pub depth: u64,
+    #[serde(rename = "type")]
+    pub ty: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct JsonRpcRequest {
+    jsonrpc: String,
+    pub method: String,
+    pub params: Params,
+    id: u32,
+}
+
+impl JsonRpcRequest {
+    fn new(method: &str, params: Params, id: u32) -> Self {
+        JsonRpcRequest {
+            jsonrpc: "2.0".to_string(),
+            method: method.to_string(),
+            params,
+            id,
+        }
+    }
+}
+#[derive(Serialize, Deserialize, Debug)]
```

Comment:
This file generally handles making the json rpc calls to the scilla server. Unf the jsonrpc rust library did not support bare tcp connections, so I did it manually. The scilla server is a bit odd in that the result is sometimes a string which contains a json object?

It was also, as far as I could tell, impossible to cleanly use this pattern below for attaching methods. This is because the backend passed to scilla has to be a reference (it is constructed in exec.rs as a reference to self). You can't even wrap it in an Arc because the compiler knows that the reference inside could potentially outlive the closure you are passing to add_method. I would be interested to hear if it is actually possible other than doing something like using channels

```
        let mut io = IoHandler::new();
        io.add_method("fetchStateValueB64", {
            let inner = Arc::clone(&inner);
            move |params| inner.lock().unwrap().fetch_state_value_b64(params)
        });
```

Reply:
Do you have a version of this in git somewhere? I'd be curious to know the error message when you use the `Arc` as in your snippet above.

Reply:
I can re-create this for you if you like but IIRC the error was something like `Arc<T> is invariant over parameter T`. Basically, the io.add_method wants to take a `fn` (not fnOnce or fnMut which are inferred based on whats moved or referred to in the closure) - this is inferred using scoping and lifetime rules.

I don't understand the mechanism too well, but I believe that `Arc` understands the properties of its generic type (its lifetime, mutability etc). This would be through subtyping and variance. If you create an Arc<&T> it understands that you could potentially get a memory violation if the closure holding the Arc clone outlives the reference to T.